src/globus_cli/login_manager/scopes.py

```diff
@@ -0,0 +1,141 @@
+from __future__ import annotations
+
+import sys
+import typing as t
+
+from globus_sdk.scopes import (
+    AuthScopes,
+    FlowsScopes,
+    GCSCollectionScopeBuilder,
+    GroupsScopes,
+    MutableScope,
+    SearchScopes,
+    TimerScopes,
+    TransferScopes,
+)
+
+from globus_cli.types import ServiceNameLiteral
+
+if sys.version_info < (3, 8):
+    from typing_extensions import Final, TypedDict
+else:
+    from typing import Final, TypedDict
+
+TRANSFER_AP_SCOPE_STR: str = (
+    "https://auth.globus.org/scopes/actions.globus.org/transfer/transfer"
+)
+
+
+def compute_timer_scope(
+    *, data_access_collection_ids: t.Sequence[str] | None = None
+) -> MutableScope:
+    transfer_scope = TransferScopes.make_mutable("all")
+    for cid in data_access_collection_ids or ():
+        transfer_scope.add_dependency(
+            GCSCollectionScopeBuilder(cid).make_mutable("data_access", optional=True)
+        )
+
+    transfer_ap_scope = MutableScope(TRANSFER_AP_SCOPE_STR)
+    transfer_ap_scope.add_dependency(transfer_scope)
+
+    timer_scope = TimerScopes.make_mutable("timer")
+    timer_scope.add_dependency(transfer_ap_scope)
+    return timer_scope
+
+
+# with no args, this builds
+#   timer[transferAP[transfer]]
+TIMER_SCOPE_WITH_DEPENDENCIES = compute_timer_scope()
+
+
+class _ServiceRequirement(TypedDict):
+    min_contract_version: int
+    resource_server: str
+    scopes: list[str | MutableScope]
+
+
+class _CLIScopeRequirements:
```

Comment:
I'll start with the note that this is purely internal, so we're free to "rearrange the furniture" as much as we like going forward, in case we don't end up resolving this fully.

I was actually thinking about going directly in the opposite direction, and making the following changes:
- rename `__getitem__` to `get_by_service_name`
- remove `__contains__`
- rename `keys` to `service_names` and return a frozenset or other well-defined collection type
- rename `values` to `__iter__` or something else? (this is the one I'm least certain about)

My reason is that I worry that it's confusing to see `value in CLI_SCOPE_REQUIREMENTS` without knowing whether `value` is expected to be a service name, resource_server string, or perhaps even a requirements dict.

I agree that the current almost-Mapping-ness is not great though. I think we either go more towards Mapping or move away from it.

Reply:
Either way works for me. As it's internal, I'm going to mash "approve" because this was more of a comment than an RFC. :grinning:

Reply:
👍 Alright, I'm going forward with this. Expect a separate PR with the change away from Mapping-ness!

Reply:
As soon as I started on that work, I _immediately_ reversed course as I saw how the results would unfold. This is going to be a subclass of `dict` with special powers. 😉